### DISCUSSION TITLE
SSH ProxyCommand

### DISCUSSION THREAD

**Discussion by Twink-Polar** [kingToolbox/WindTerm#36] [General]
Would you consider adding the ability for windTerm to use the SSH ProxyCommand?

very handy for accessing a host via an intermediary box

**Comment by kingToolbox** [comment]
Sorry for the late reply. Yes, this has already on the todo list of [v2.0](https://github.com/kingtoolbox/windterm/#roadmap). But it may only support the Linux version, and it is a little difficult to implement this on the Windows version.

In addition, maybe you can try "Execute command after authentication" (Session's property dialog / SSH / SSH / Authentication), set it like "ssh root@10.0.0.1",  if you only jump once, then this is similar.

**Reply by Twink-Polar** [reply]
I guess the excute cmd after auth wouldn’t work if you were using nc (netcat) ? An additional use case I have is for tunnelling ssh over tor. Really useful when coupled with a hidden service.

**Reply by kingToolbox** [reply]
"Execute command after authentication" is designed to execute some simple commands, but it can also execute scripts like "sh ./init.sh", so it should be able to do very complicated things by calling scripts.

Of course, this cannot replace ProxyCommand. If everything goes well, ProxyCommand will be implemented in the next version. I hope to implement this function in both Linux and Windows.

I’m busy improving [WindEdit](https://github.com/kingToolbox/digedit/) recently, which supports TB file, GB line and billions lines. So WindEdit v1.0 will be released in early January, and WindTerm v2.0 is expected to be released in mid-to-late January.

**Comment by Twink-Polar** [comment]
Just been test driving the 2.0 pre-release. Very nice. 

I see that you pull out of the ssh config (if present) hostname and user etc but wondering if you are planning on doing the same for the ProxyCommand?

**Reply by kingToolbox** [reply]
Yes, I will definitely do this, it is very convenient for users to migrate from OpenSSH. In addition, if my pre-research is correct, the windows version will also realize ProxyCommand. I think I have found a good way to achieve it.

The v2.0.0_prerelease_1 is just the beginning, there are still many new important features waiting to be merged, including a new memory manager and garbage collector, search text in sessions, and of course ProxyCommand and so on. The completed version will be able to hold billions of lines of text and will only take up very little memory.

I still look forward to release v2.0.0 in late January.

**Reply by Twink-Polar** [reply]
Can't wait!

**Comment by Twink-Polar** [comment]
Hi,

How goes the 2.0 development?

**Reply by kingToolbox** [reply]
Sorry for the late reply. I have been busy preparing for [v2.0.0_Prerelease_2](https://github.com/kingToolbox/WindTerm/releases/tag/2.0.0) during this period. As I said in the previous post, the newly completed memory manager and garbage collector are able to hold billions of lines of text by using a small amount of memory. Not only the memory footprint is greatly reduced, but the speed is also faster, which will lay a good foundation for the development of the new v2.x series.

The next step is to implement the ProxyCommand, which is the last piece of SSH protocol in WindTerm. I am still thinking about how to implement automatic login for multiple jump servers, if they use password authentication instead of public key authentication. I expect to complete this feature early next week, and I will release v2.0.0_Prerelease_3 at that time. Welcome to download and put forward valuable suggestions.

The v2.0.0 stable version may be delayed for a while, I expect it to be released in mid-February. And search text in sessions has been re-planned and postponed to the v2.1.0 now.

**Reply by Twink-Polar** [reply]
Thank you for the update. I'll download pre-release 2 and take it for a spin.

**Comment by kingToolbox** [comment]
The v2.0.0 stable version has been finally released and ProxyCommand also has been implemented, please download and use it. And you can read intros [ProxyCommand](https://kingtoolbox.github.io/2021/03/11/proxycommand/), [Network Proxy / Jump Server](https://kingtoolbox.github.io/2021/03/11/proxy-jump-server/) and [Network Proxy / HTTP and SOCKS5 proxy](https://kingtoolbox.github.io/2021/03/11/proxy-http-socks5/) to know more. I believe the current design is very easy to use. If there are any needs, please let me know. Thank you!
